### PR TITLE
Add SPI Library

### LINKED ISSUE #87: Implementing SPI library
We need to implement the SPI library to keep it compatible with standard AVR Arduino.

### PR DESCRIPTION
**Pull request by Ayush1325**
- Tested master mode on beagleconnect freedom.
- Add master sample.
- Interrupts not implemented. Slave mode not tested.
- Fixes #87

### CODE REVIEW

File: libraries/SPI/SPI.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2024 Ayush Singh <ayush@beagleboard.org>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "SPI.h"
+#include <zephyr/kernel.h>
+
+uint8_t SPCR = 0;
```
Comment (DhruvaG2000) [inline libraries/SPI/SPI.cpp]: No need of ``= 0`` IMO
See https://stackoverflow.com/questions/16015656/are-global-variables-always-initialized-to-zero-in-c

Comment (DhruvaG2000) [reply libraries/SPI/SPI.cpp]: Also, may want to add a comment at the top saying
```
/* Serial Peripheral Control Register */
```
As some people may even call it SPIGCR or something else. 
See https://ti.com/lit/ug/sprugp2a/sprugp2a for eg.

File: libraries/SPI/SPI.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  const struct spi_buf_set tx_buf_set = {
+      .buffers = &tx_buf,
+      .count = 1,
+  };
+  const struct spi_buf rx_buf = {.buf = &rx, .len = sizeof(rx)};
+  const struct spi_buf_set rx_buf_set = {
+      .buffers = &rx_buf,
+      .count = 1,
+  };
+  int ret;
```
Comment (DhruvaG2000) [inline libraries/SPI/SPI.cpp]: No mixed declarations please.
FYI: https://langdev.stackexchange.com/questions/3115/why-are-mixed-declarations-more-challenging-to-implement-than-forcing-all-declar

File: libraries/SPI/SPI.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+
```
Comment (DhruvaG2000) [inline libraries/SPI/SPI.cpp:29]: remove extra blank lines

File: samples/spi_master/README.rst [hunk 4]
```diff
@@ -0,0 +1,9 @@
+.. _hello_world:
+
+SPI Master
```
Comment (DhruvaG2000) [inline samples/spi_master/README.rst]: Please make use of more inclusive terminology.  This should be addressed throughout this PR.
Read https://www.oshwa.org/a-resolution-to-redefine-spi-signal-names/

SPI controller makes more sense.

File: samples/spi_master/src/app.cpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2024 Ayush Singh <ayush@beagleboard.org>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "SPI.h"
+#include <Arduino.h>
+
+#define SS 3
```
Comment (DhruvaG2000) [inline samples/spi_master/src/app.cpp]: Very ambiguous name, use something more verbose

**Review by DhruvaG2000 — CHANGES_REQUESTED** [review]

File: samples/spi_master/src/app.cpp (hunk 5 above)
Comment (Ayush1325) [reply samples/spi_master/src/app.cpp]: So, the reason for choosing `SS` is that it seems to be defined in standard arduino as the CS Pin.

File: libraries/SPI/SPI.cpp (hunk 1 above)
Comment (Ayush1325) [reply libraries/SPI/SPI.cpp]: @DhruvaG2000 the name is taken from [Arduino SPI docs](https://docs.arduino.cc/tutorials/generic/introduction-to-the-serial-peripheral-interface/). So I think we should stick with it?

File: libraries/SPI/SPI.cpp (hunk 2 above)
Comment (Ayush1325) [reply libraries/SPI/SPI.cpp]: @DhruvaG2000 Do you just mean moving `ret` at the top or remove mixed declaration (like C89)? That would require making everything non-const.

Comment (DhruvaG2000) [reply libraries/SPI/SPI.cpp]: just move it to the top.

File: libraries/SPI/SPI.cpp (hunk 1 above)
Comment (DhruvaG2000) [reply libraries/SPI/SPI.cpp]: @Ayush1325  My comment wasn't on the name, I'm just saying write the full form as a comment above

File: samples/spi_master/src/app.cpp (hunk 5 above)
Comment (DhruvaG2000) [reply samples/spi_master/src/app.cpp]: https://docs.arduino.cc/tutorials/generic/introduction-to-the-serial-peripheral-interface/

chip sel/ CS makes more sense.

Comment (DhruvaG2000) [reply samples/spi_master/src/app.cpp]: > Controller/peripheral is formerly known as master/slave. Arduino no longer supports the use of this terminology. See the table below to understand the new terminology

**Review by DhruvaG2000 — APPROVED** [review]
I think we'll also have to add the spi node to add support for other platforms as well, but I don't suppose you'd have the hw to test it all anyway.
So I am good with this PR, thanks for addressing all concerns!

**Pull request merged**